pkg/retagger/executor.go

```diff
@@ -21,7 +29,148 @@ func (r *Retagger) ExecuteJobs() error {
 	return nil
 }
 
+// executeJob accepts a job definition and runs it as either a pattern job or a single job
 func (r *Retagger) executeJob(job Job) error {
+	if job.SourceSha != "" {
+		if job.SourcePattern != "" {
+			// Configuration specified a SHA and a pattern -- use SHA to be safe, but warn about misconfiguration
+			r.logger.Log("level", "warn", "message", fmt.Sprintf("invalid configuration: Job %v specifies both a SHA (%v) and a Pattern (%v). Using SHA", job.SourceImage, job.SourceSha, job.SourcePattern))
+		}
+		return r.executeSingleJob(job, true) // Skip existing tags for this job
+	}
+	return r.executePatternJob(job)
+}
+
+// executePatternJob accepts a job definition containing a tag pattern
+// and creates and runs single jobs for each tag matching the pattern.
+func (r *Retagger) executePatternJob(job Job) error {
+	jobs, err := r.compilePatternJobs(job)
+	if err != nil {
+		return err
+	}
+
+	for _, job := range jobs {
+		r.executeSingleJob(job, false) // Do not skip existing tags for this job
+	}
+
+	return nil
+}
+
+// compilePatternJobs gets tag lists from the upstream repo, finds tags matching the pattern specified in the job,
+// and creates new jobs to retag new images and images which have changed since their last collection.
+func (r *Retagger) compilePatternJobs(job Job) ([]Job, error) {
+	r.logger.Log("level", "debug", "message", fmt.Sprintf("compiling jobs for image %v using pattern %v, with options %#v", job.SourceImage, job.SourcePattern, job.Options))
+
+	// Make sure our pattern is valid
```

Comment:
```suggestion
	// Make sure our pattern is valid.
```

You get the _point_ for the rest of these :P

Reply:
_Point_ taken